crates/stateless/src/validation.rs

```diff
@@ -0,0 +1,292 @@
+
+    // Execute the block
+    let basic_block_executor = EthExecutorProvider::ethereum(chain_spec.clone());
+    let executor = basic_block_executor.executor(db);
+    let output = executor
+        .execute(&current_block)
+        .map_err(|e| StatelessValidationError::StatelessExecutionFailed(e.to_string()))?;
+
+    // Post validation checks
+    validate_block_post_execution(&current_block, &chain_spec, &output.receipts, &output.requests)
+        .map_err(|err| StatelessValidationError::StatelessPostValidationFailed(err.to_string()))?;
+
+    // Compute and check the post state root
+    // TODO: Remove rayon
+    let hashed_state =
+        HashedPostState::from_bundle_state::<KeccakKeyHasher>(output.state.state.par_iter());
+    let state_root = crate::root::calculate_state_root(&mut sparse_trie, hashed_state)
+        .map_err(|_e| StatelessValidationError::StatelessStateRootCalculationFailed)?;
+    if state_root != current_block.state_root {
+        return Err(StatelessValidationError::PostStateRootMismatch {
+            got: state_root,
+            expected: current_block.state_root,
+        });
+    }
+
+    // Return block hash
+    Ok(current_block.hash_slow())
+}
+
+/// Verifies execution witness [`ExecutionWitness`] against an expected pre-state root.
+///
+/// This function takes the RLP-encoded values provided in [`ExecutionWitness`]
+/// (which includes state trie nodes, storage trie nodes, and contract bytecode)
+/// and uses it to populate a new [`SparseStateTrie`].
+///
+/// If the computed root hash matches the `pre_state_root`, it signifies that the
+/// provided execution witness is consistent with that pre-state root. In this case, the function
+/// returns the populated [`SparseStateTrie`] and a [`B256Map`] containing the
+/// contract bytecode (mapping code hash to [`Bytecode`]).
+///
+/// The bytecode has a separate mapping because the [`SparseStateTrie`] does not store the
+/// contract bytecode, only the hash of it (code hash).
+///
+/// If the roots do not match, it returns `None`, indicating the witness is invalid
+/// for the given `pre_state_root`.
+// Note: This approach might be inefficient for ZKVMs requiring minimal memory operations, which
+// would explain why they have for the most part re-implemented this function.
+pub fn verify_execution_witness(
+    witness: &ExecutionWitness,
+    pre_state_root: B256,
+) -> Result<(SparseStateTrie, B256Map<Bytecode>), StatelessValidationError> {
+    let mut trie = SparseStateTrie::new(DefaultBlindedProviderFactory);
+    let mut state_witness = B256Map::default();
+    let mut bytecode = B256Map::default();
+
+    for rlp_encoded in &witness.state {
+        let hash = keccak256(rlp_encoded);
+        state_witness.insert(hash, rlp_encoded.clone());
+    }
+    for rlp_encoded in &witness.codes {
+        let hash = keccak256(rlp_encoded);
+        bytecode.insert(hash, Bytecode::new_raw(rlp_encoded.clone()));
+    }
```

Comment:
https://github.com/paradigmxyz/reth/pull/15591/files#r2040307201

same can be said about the bytecodes. there is no validation that all the bytecodes have corresponding leaves in the witness